system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1331,354 @@ handle_run_triggers (FlatpakSystemHelper   *object,
+                                             "/org/freedesktop/DBus",
+                                             "org.freedesktop.DBus",
+                                             "GetConnectionCredentials",
+                                             g_variant_new ("(s)", sender),
+                                             G_VARIANT_TYPE ("(a{sv})"), G_DBUS_CALL_FLAGS_NONE, G_MAXINT, NULL, error);
+  if (credentials == NULL)
+    return FALSE;
+
+   dict = g_variant_get_child_value (credentials, 0);
+  unixUserId = g_variant_lookup_value (dict, "UnixUserID", G_VARIANT_TYPE_UINT32);
+
+   if (unixUserId == NULL)
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                   "Failed to query UnixUserId for the bus name: %s", sender);
+      return FALSE;
+    }
+
+   *uid_out = g_variant_get_uint32 (unixUserId);
+  return TRUE;
+}
+
+
+static void
+revokefs_child_setup (gpointer user_data)
+{
+  struct passwd *passwd = getpwnam("flatpak");
+  setgid (passwd->pw_gid);
+  setuid (passwd->pw_uid);
+
+  close (GPOINTER_TO_INT(user_data));
+}
+
+static void
+name_vanished_cb (GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+  OngoingPull *pull = (OngoingPull *) user_data;
+
+  ongoing_pull_free (pull);
+}
+
+static OngoingPull *
+ongoing_pull_new (FlatpakSystemHelper   *object,
+                  GDBusMethodInvocation *invocation,
+                  const gchar           *ref,
+                  const gchar           *commit,
+                  uid_t                  uid,
+                  gchar                 *revokefs_fuse_path,
+                  gchar                 *src,
+                  gchar                 *mnt,
+                  GError               **error)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  OngoingPull *pull;
+  g_autoptr (GSubprocessLauncher) launcher = NULL;
+  g_autoptr (GSubprocess) revokefs_backend = NULL;
+  g_autofree gchar *target_socket = NULL;
+  int sockets[2];
+
+  pull = g_slice_new0 (OngoingPull);
+  pull->object = object;
+  pull->invocation = invocation;
+  pull->ref = g_strdup (ref);
+  pull->commit = g_strdup (commit);
+  pull->src_dir = g_strdup (src);
+  pull->mountpoint = g_strdup (mnt);
+  pull->cancellable = g_cancellable_new ();
+  pull->uid = uid;
+  pull->preserve_pull = FALSE;
+
+  pull->watch_id = g_bus_watch_name_on_connection (connection,
+                                                   g_dbus_connection_get_unique_name (connection),
+                                                   G_BUS_NAME_WATCHER_FLAGS_NONE, NULL,
+                                                   name_vanished_cb,
+                                                   pull, NULL); /* TODO: Take a ref here */
+
+  if (socketpair (AF_UNIX, SOCK_SEQPACKET, 0, sockets) == -1)
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED, "Syscall socketpair failed");
+      return NULL;
```

Comment:
Leaks `pull`

Reply:
Not really. If `NULL` is returned by `ongoing_pull_new`, it is free-d in the error handling. See where `ongoing_pull_new` is called.

Reply:
I don't agree. Look at the code: ongoing_pull_new() calls `g_slice_new0` and assigns the result to `pull`. It fills in a bunch of fields. Then it calls `socketpair`: if that fails, it returns `NULL` to the caller. How can the caller clean up the memory allocated?